### PR TITLE
fix(aws-ssm): correctly emit Association.Name

### LINKED ISSUE #852: Framework: remove property renames for CloudFormation constructs
in `AssociationResourceProps`: documentName is `documentName`
in the CF template, this becomes `DocumentName`

This fails when deploying the stack. According to the [documentation](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ssm-association.html), the parameter should be `Name`. Manually modifying the CF template generated by CDK to use the correct name works.

**My workaround**
Because `AssociationResource` is a L1 model, I have built a wrapper around it to do what I need. I have change my wrapper to instead create an instance of OverloadedAssociation - a class I created that extends AssociationResource. 

```
  /**
   * ``AssociationResource`` does not generate the correct CF file, one of the parameters needs to be renamed.
   */
  class OverloadedAssociation extends Ssm.AssociationResource {
    /**
     * Creates a new ``AWS::SSM::Association``.
     *
     * @param parent   the ``cdk.Construct`` this ``AssociationResource`` is a part of
     * @param name     the name of the resource in the ``cdk.Construct`` tree
     * @param properties the properties of this ``AssociationResource``
     */
    constructor(parent: Construct, name: string, properties: Ssm.AssociationResourceProps) {
      super(parent, name, properties);
    }

    protected renderProperties(properties: any): { [key: string]: any } {
      const baseResult = super.renderProperties(properties);

      return {
        // this is the line that changed
        Name: baseResult.DocumentName,
        AssociationName: baseResult.AssociationName,
        DocumentVersion: baseResult.DocumentVersion,
        InstanceId: baseResult.InstanceId,
        OutputLocation: baseResult.OutputLocation,
        Parameters: baseResult.Parameters,
        ScheduleExpression: baseResult.ScheduleExpression,
        Targets: baseResult.Targets,
      };
    }
  }
```

Is this the right way to workaround this problem? Should I be modifying `.ToCloudFormation()` instead?

### PR DESCRIPTION
**Pull request by rix0rrr**
The rename used to happen in the cfnspec, which helped the model
generate correct names, but the code generator would lose the ability to
access the original name, and so it could not be emitted correctly.

Add a property renaming feature to the 'cfn2ts' code generator
(configured via a special entry in package.json).

Fixes #852.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

### CODE REVIEW

File: tools/cfn2ts/lib/codegen.ts [hunk 1]
```diff
@@ -143,28 +146,37 @@ export default class CodeGenerator {
...
+        propName = resourceNameProperty(resourceName);
+      }
+
+      if (originalName !== propName) {
         // tslint:disable-next-line:no-console
-        console.error(`Renamed property 'Name' of ${resourceName.fqn} to '${renamedNameProperty}'`);
+        console.error(`[${resourceName.fqn}] Renamed property '${originalName}' to '${propName}'`);
       }
+      if (finalNames.has(propName)) {
+        throw new Error(`Oh gosh, there was already a property named ${propName}!`);
```
Comment (eladb) [inline tools/cfn2ts/lib/codegen.ts]: Thanks for preserving the spirit

**Comment by eladb** [comment]
Not sure I like this and it makes me wonder: maybe we should just get rid of the "name" => "xxxName" renames altogether in L1. The reason we are actually doing these renames was that the construct ID used to be a "name" prop, so if an L1 had a `Name` property, they would conflict. 

However, now construct IDs are not a prop, and we technically don't need these renames at all.

**Comment by rix0rrr** [comment]
>  The reason we are actually doing these renames was that the construct ID used to be a "name" prop, so if an L1 had a `Name` property, they would conflict.

True, that was the original reason. At the same time, I do like the transformation we're doing right now, if only to make the object model more consistent. The CloudFormation object model isn't always consistent in the names it uses, but I'd rather have our model be consistent.

For example (in no particular order, just random sampling from the CloudFormation model):

* **AWS::Kinesis::Stream**, **AWS::CodeCommit::Repository**: property **Name**.
* **AWS::CodeDeploy::Application**, property **ApplicationName**, **AWS::DynamoDB::Table**, property **TableName**.
* Not to mention **AWS::SSM::Association** which has both **Name** and **AssociationName**.

And at the same time, when we generate `{GetAtt}` and `{Ref}` getters, we do always add in the resource type into the name (for example `{ Fn::GetAtt: [Bucket, Arn] }` -> `bucket.bucketArn` (as opposed to `bucket.arn`)). For consistency's sake we do use the resource type in there a lot, and I don't mind the explicit verbosity here.

I'm aware we could be providing this uniformity at the L2 level as well... but this is just such a convenient place to add it. I'm torn.

**Comment by eladb** [comment]
I think there is value in staying 1:1 with the CFN resource spec at L1. It makes it easier for people to navigate this layer, find help and in general have less magic. I vote for removing this sugar...

**Comment by RomainMuller** [comment]
Regarding the `name` prop... There are other conflict cases that we very much ignore at the moment, though... What will happen when some CFN resource gets a `Path` attribute is going to be not pretty.

**Comment by eladb** [comment]
@RomainMuller This is about the resource properties (`props`) not attributes, which are implemented as the properties of the construct (and we prefix the resource name to them all).

**Comment by eladb** [comment]
@rix0rrr do you want me to pick this up?

**Pull request closed without merging**